### PR TITLE
ci(tofu-ci): tf tests

### PR DESCRIPTION
**Pull request by rdhar**

### CODE REVIEW

**Comment by chris3ware** [comment]
Hi, thanks for taking the time to suggest improvements to the workflow. I am trying to run plan / apply then another plan to check for diffs and finally destroy to check for any dependancies in the development environment. Then run the same again, minus the destroy, in my mock production environment. My first attempt did try and use your action for this, but it failed: https://github.com/3ware/aws-network-speciality/actions/runs/10833169395/job/30059298723 - which is why I was experimenting with `tofu` commands.

File: .github/workflows/tofu-ci.yaml [hunk 1]
```diff
@@ -24,52 +24,53 @@ jobs:
...
-      backend-credentials: ${{ secrets.TF_TOKEN_APP_TERRAFORM_IO }}
-      provider-credentials: ${{ secrets.AWS_DEV_OIDC_ROLE_ARN }}
 
-    #TODO Copy files from dev to prod
+    strategy:
+      matrix:
+        terraform-dir: ${{ fromJSON(needs.find-terraform-dir.outputs.terraform-dir) }}
+
+    # Set environment only
+    environment: ${{ github.event_name == 'push' && 'development' || '' }}
```
Comment (chris3ware) [inline .github/workflows/tofu-ci.yaml:46]: Please can you explain what this does?

**Review by chris3ware — COMMENTED** [review]
Please can you confirm that these changes are just for the initial plan and apply, and that another job is required to run plan again to check for a diff?

**Comment by rdhar** [comment]
It might help to share a "decision tree" of sorts to understand what chain of actions should take place under which condition. To start with, let's tackle the dev env for example:

```mermaid
graph TD;
    A[PR opened] --> C[TF plan]
    B[PR updated] --> C[TF plan] --> D[PR merged]
    D[PR merged] --> E[TF apply] --> F[TF destroy]
```

Would you mind clarifying the reason for the second `plan` — why check for diffs immediately after `apply`? If `apply` is successful, then there's no diff, otherwise the `apply` failed.

And am I right in think you're wanting to `destroy -auto-approve` immediately after to deprovision any/all resources?

**Comment by chris3ware** [comment]
> It might help to share a "decision tree" of sorts to understand what chain of actions should take place under which condition. To start with, let's tackle the dev env for example:
> 
> ```mermaid
> graph TD;
>     A[PR opened] --> C[TF plan]
>     B[PR updated] --> C[TF plan] --> D[PR merged]
>     D[PR merged] --> E[TF apply] --> F[TF destroy]
> ```
> Would you mind clarifying the reason for the second `plan` — why check for diffs immediately after `apply`? If `apply` is successful, then there's no diff, otherwise the `apply` failed.
> 
> And am I right in think you're wanting to `destroy -auto-approve` immediately after to deprovision any/all resources?

The reason for the second plan is to check idempotence of the apply. Sometimes applies can be successful, but another plan will reveal a drift or similar.

That's correct regarding `destroy`. The destroy is check the apply has not introduced any unexpected dependancies.

**Comment by rdhar** [comment]
Right, I see, and what's the expected outcomes of the idempotency check?
- `0` success (no changes) > clean up.
- `1` error > end.
- `2` success (with diff) > ?

Here's the updated decision tree / flow chart, as best I understand it.

```mermaid
flowchart TD
    A[PR opened] -->|TF plan| C(PR merged)
    B[PR updated] -->|TF plan| C
    C -->|TF apply| D{Exit code}
    D -->|0-success| E[Idempotence] -->|TF plan| F{Exit code}
    F -->|0-success no-changes| G[Clean up] -->|TF destroy| H(End)
    F -->|1-error| H
    F -->|2-success with-diff| H
    D -->|1-error| H
```

**Comment by rdhar** [comment]
> Please can you confirm that these changes are just for the initial plan and apply, and that another job is required to run plan again to check for a diff?

Updated this PR to plan/apply infrastructure, followed by another plan with exitcode with idempotency check, and finally clean up all resources — all in the one job with just a few steps.

**Comment by chris3ware** [comment]
> Right, I see, and what's the expected outcomes of the idempotency check?
> 
> * `0` success (no changes) > clean up.
> * `1` error > end.
> * `2` success (with diff) > ?
> 

I guess, for `2`. an issue could be raised to investigate.

**Comment by chris3ware** [comment]
> > Please can you confirm that these changes are just for the initial plan and apply, and that another job is required to run plan again to check for a diff?
> 
> Updated this PR to plan/apply infrastructure, followed by another plan with exitcode for idempotency check, and conditionally clean up all resources — all in the one job with just a few steps.

This is great. Thanks for your input.

**Pull request merged**